### PR TITLE
CMake adopt to latest eager runtime and reusable shared library in windows environment

### PR DESCRIPTION
**Pull request by jackyko1991**
Due to some error in CLA checking https://github.com/tensorflow/tensorflow/pull/16394, a new pull request is made

I am currently working on the cmake files to generate shared library that allows users to develop with C++ interface.

It is easier to develop by adding export targets in tf_shared_lib.cmake, but I think in the future this function should be migrated to the main CMakeList.txt

### CODE REVIEW

**Comment by jackyko1991** [comment]
Accompany with this pull request with the following CMakeLists for C++ api hello world.

```cmake
cmake_minimum_required (VERSION 2.6)
project (tf_hello)

# Tensorflow
find_package(Tensorflow REQUIRED)
include_directories(${TENSORFLOW_INCLUDE_DIRS})
message(STATUS ${TENSORFLOW_INCLUDE_DIRS})

# compiler setting required by tensorflow, to be test on all compilers
# currently only tested on MSVC and GCC
if (${CMAKE_CXX_COMPILER_ID} STREQUAL MSVC) 
	add_definitions(-DCOMPILER_MSVC)
elseif (${CMAKE_CXX_COMPILER_ID} STREQUAL GNU)
	if (${CMAKE_CXX_COMPILER_VERSION} VERSION_LESS "3")
		add_definitions(-DCOMPILER_GCC3)
	else()
		add_definitions(-D__GNUC__)
	endif()
else()
	message(ERROR " compiler ${CMAKE_CXX_COMPILER_ID} not supported by this CMakeList.txt, under development")
endif()

add_executable(tf_hello main.cxx)
target_link_libraries(tf_hello ${TENSORFLOW_LIBRARIES})

message(STATUS ${TENSORFLOW_LIBRARIES})
```

main.cxx can be found from https://www.tensorflow.org/api_guides/cc/guide

File: tensorflow/core/platform/default/gpu/cupti_wrapper.h [hunk 1]
```diff
@@ -20,10 +20,10 @@ limitations under the License.
 
 #include <stddef.h>
 #include <stdint.h>
-#if defined(WIN32)
-#include "extras/CUPTI/include/cupti.h"
-#else
+#if defined(PLATFORM_GOOGLE)
```
Comment (jackyko1991) [inline tensorflow/core/platform/default/gpu/cupti_wrapper.h]: I am not very sure that this is good way to set correct platform. In typical cuda usage, the prepending cuda folder is not necessary since the cuda SDK is setted during dependency configuration.

File: tensorflow/core/util/cuda_device_functions.h [hunk 2]
```diff
@@ -28,12 +28,21 @@ limitations under the License.
 
 #include <algorithm>
 #include <complex>
+#ifdef PLATFORM_GOOGLE
```
Comment (jackyko1991) [inline tensorflow/core/util/cuda_device_functions.h]: I am not familiar with bazel build, but this change get build pass with cmake.

**Comment by jackyko1991** [comment]
After updating protobuf to 3.5, it request NASM, perl and go for grpc compilation, how can I add these compilers to the check build environment?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewers @yifeif, @mrry: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by apattnaik0721013** [comment]
@jackyko1991 can you fix this conflict?

**Comment by jackyko1991** [comment]
@apattnaik0721013 conflict solved

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@
 cmake_minimum_required(VERSION 3.5)
 
 # Project
-project(tensorflow C CXX)
+project(tensorflow VERSION 1.7.0 LANGUAGES C CXX)
```
Comment (mrry) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Please update this from 1.7.0 to 1.8.0.

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 4]
```diff
@@ -63,6 +69,11 @@ if (NOT WIN32)
     # option's default value is OFF. Fill it with real default values
     set(tensorflow_CUDNN_INCLUDE /usr/include)
   endif (NOT tensorflow_CUDNN_INCLUDE)
+  option(tensorflow_NCCL_INCLUDE "nccl.h headher install path" /usr/include/)
```
Comment (mrry) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Nit: Replace "headher" with "header".

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 5]
```diff
@@ -155,9 +166,6 @@ if(WIN32)
   set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /D_ITERATOR_DEBUG_LEVEL=0")
   set(CMAKE_CXX_FLAGS_MINSIZEREL "${CMAKE_CXX_FLAGS_MINSIZEREL} /D_ITERATOR_DEBUG_LEVEL=0")
   set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /D_ITERATOR_DEBUG_LEVEL=0")
-
-  # Try to avoid flaky failures due to failed generation of generate.stamp files.
-  set(CMAKE_SUPPRESS_REGENERATION ON)
```
Comment (mrry) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Please add these lines back, because they are necessary to make the build more robust on some platforms.

File: tensorflow/contrib/cmake/README.md [hunk 6]
```diff
@@ -2,9 +2,9 @@ TensorFlow CMake build
 ======================
 
 This directory contains CMake files for building TensorFlow on Microsoft
-Windows. [CMake](https://cmake.org) is a cross-platform tool that can
+Windows and Linux. [CMake](https://cmake.org) is a cross-platform tool that can
 generate build scripts for multiple build systems, including Microsoft
-Visual Studio.
+Visual Studio and GCC. The method suppose could work on MacOS, but not tested yet.
```
Comment (mrry) [inline tensorflow/contrib/cmake/README.md]: Replace last sentence with: "The method has not been tested on Mac OS X."

File: tensorflow/contrib/cmake/README.md [hunk 7]
```diff
@@ -13,12 +13,13 @@ Linux.
 Current Status
 --------------
 
-CMake can be used to build TensorFlow on Windows. See the [getting started documentation](https://www.tensorflow.org/install/install_windows)
-for instructions on how to install a pre-built TensorFlow package on Windows.
+CMake can be used to build TensorFlow on all platforms. See the [getting started documentation](https://www.tensorflow.org/install/install_windows)
+for instructions on how to install a pre-built TensorFlow package on Windows and Linux. Procedure in MacOS is similar to the Linux build.
```
Comment (mrry) [inline tensorflow/contrib/cmake/README.md]: Replace "Procedure" with "The procedure"

File: tensorflow/contrib/cmake/README.md [hunk 8]
```diff
@@ -13,12 +13,13 @@ Linux.
...
 
 ### Current known limitations
 * It is not possible to load a custom Op library.
 * GCS file system is not supported.
+* Debug build is not available since python no longer release debug library.
```
Comment (mrry) [inline tensorflow/contrib/cmake/README.md]: Replace "since python no longer release debug library." with "since Python for Windows is no longer distributed with a debug library."

File: tensorflow/contrib/cmake/external/grpc.cmake [hunk 9]
```diff
@@ -20,24 +20,17 @@ set(GRPC_BUILD ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc)
...
-  endif()
+  set(grpc_STATIC_LIBRARIES
+      ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/$(Configuration)/grpc++_unsecure.lib
+      ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/$(Configuration)/grpc_unsecure.lib
+      ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/$(Configuration)/gpr.lib)
 else()
   set(grpc_STATIC_LIBRARIES
       ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/libgrpc++_unsecure.a
       ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/libgrpc_unsecure.a
-      ${CMAKE_CURRENT_BINARY_DIR}/grpc/src/grpc/libaddress_sorting.a
```
Comment (mrry) [inline tensorflow/contrib/cmake/external/grpc.cmake]: Is `libaddress_sorting.a` no longer required on Linux?

File: tensorflow/contrib/cmake/external/grpc.cmake [hunk 10]
```diff
@@ -59,7 +52,9 @@ ExternalProject_Add(grpc
         -DPROTOBUF_INCLUDE_DIRS:STRING=${PROTOBUF_INCLUDE_DIRS}
         -DPROTOBUF_LIBRARIES:STRING=${protobuf_STATIC_LIBRARIES}
         -DZLIB_ROOT:STRING=${ZLIB_INSTALL}
-	-DgRPC_SSL_PROVIDER:STRING=NONE
+        -DgRPC_SSL_PROVIDER:STRING=NONE
+        # -DCMAKE_ASM_NASM_COMPILER:STRING=${CMAKE_ASM_NASM_COMPILER}
```
Comment (mrry) [inline tensorflow/contrib/cmake/external/grpc.cmake]: Delete these commented out lines.

File: tensorflow/contrib/cmake/external/protobuf.cmake [hunk 11]
```diff
@@ -16,7 +16,16 @@ include (ExternalProject)
 
 set(PROTOBUF_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/protobuf/src/protobuf/src)
 set(PROTOBUF_URL https://github.com/google/protobuf.git)
-set(PROTOBUF_TAG b04e5cba356212e4e8c66c61bbe0c3a20537c5b9)
+
+# enable choose protobuf versions
```
Comment (mrry) [inline tensorflow/contrib/cmake/external/protobuf.cmake]: I think there is a build issue with the latest version of Protobuf on Windows. @av8ramit Do you know if this is true or if it's been resolved?

File: tensorflow/contrib/cmake/tf_python.cmake [hunk 12]
```diff
@@ -422,8 +429,8 @@ GENERATE_PYTHON_OP_LIB("contrib_text_skip_gram_ops"
   DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/tf_python/tensorflow/contrib/text/python/ops/gen_skip_gram_ops.py)
 GENERATE_PYTHON_OP_LIB("contrib_bigquery_reader_ops"
   DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/tf_python/tensorflow/contrib/cloud/python/ops/gen_bigquery_reader_ops.py)
-GENERATE_PYTHON_OP_LIB("stateless_random_ops"
-  DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/tf_python/tensorflow/contrib/stateless/gen_stateless_random_ops.py)
+# GENERATE_PYTHON_OP_LIB("stateless_random_ops"
```
Comment (mrry) [inline tensorflow/contrib/cmake/tf_python.cmake]: Uncomment these lines.

File: tensorflow/core/platform/default/logging.h [hunk 13]
```diff
@@ -187,6 +187,11 @@ string* MakeCheckOpString(const T1& v1, const T2& v2, const char* exprtext) {
   return comb.NewString();
 }
 
+#ifdef COMPILER_MSVC
```
Comment (mrry) [inline tensorflow/core/platform/default/logging.h]: Can this move to `tensorflow/core/platform/macros.h`?

File: tensorflow/core/platform/default/logging.h [hunk 14]
```diff
@@ -187,6 +187,11 @@ string* MakeCheckOpString(const T1& v1, const T2& v2, const char* exprtext) {
...
+#undef max;
+#undef min;
```
Comment (mrry) [inline tensorflow/core/platform/default/logging.h]: Delete the semicolons at the end of these lines.

File: tensorflow/core/util/cuda_device_functions.h [hunk 15]
```diff
@@ -28,10 +28,23 @@ limitations under the License.
 
 #include <algorithm>
 #include <complex>
-#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
```
Comment (mrry) [inline tensorflow/core/util/cuda_device_functions.h]: Why is this line deleted?

File: tensorflow/core/util/cuda_device_functions.h [hunk 16]
```diff
@@ -28,10 +28,23 @@ limitations under the License.
...
+#ifdef PLATFORM_GOOGLE
```
Comment (mrry) [inline tensorflow/core/util/cuda_device_functions.h]: This change seems like it will break non-`PLATFORM_GOOGLE` builds, such as the Linux-with-Bazel GPU build.

File: tensorflow/tools/pip_package/setup.py [hunk 17]
```diff
@@ -39,9 +40,10 @@
     'gast >= 0.2.0',
     'numpy >= 1.13.3',
     'six >= 1.10.0',
-    'protobuf >= 3.4.0',
+    'protobuf >= 3.5.0',
```
Comment (mrry) [inline tensorflow/tools/pip_package/setup.py]: @martinwicke Is there a particular person in charge of approving PIP dependency changes, and can you please loop them in to this change?

**Review by mrry — CHANGES_REQUESTED** [review]

**Comment by jackyko1991** [comment]
@mrry Code modified according to review

**Comment by Skydes** [comment]
Would it be possible to have a better control over the exported symbols ? Currently we have a symbol conflict with OpenCV (because of libpng and libjpeg, see https://github.com/tensorflow/tensorflow/issues/14267).

Could we also work towards having GRPC optional on Linux ? The main issue would be with [tensorflow/core/debug/debug_io_utils.h](../blob/master/tensorflow/core/debug/debug_io_utils.cc)

File: tensorflow/tools/pip_package/setup.py (hunk 17 above)
Comment (martinwicke) [reply tensorflow/tools/pip_package/setup.py]: If it passes the tests and it's available for all platforms then this is fine.

**Comment by martinwicke** [comment]
It would be great to compile with all symbols hidden except those that need to be public. @allenlavoie, @gunan

**Comment by allenlavoie** [comment]
@Skydes @martinewicke We already run a linker script for the C++ and C APIs on other platforms (we should do it here too). The issue with https://github.com/tensorflow/tensorflow/issues/14267 is that the C++ API doesn't include protocol buffer symbols, and to get those you need to link against something that needs to export the conflicting symbols (I've [outlined a solution](https://github.com/tensorflow/tensorflow/issues/9525#issuecomment-385475293) if you're interested, although it's probably not relevant to this PR since custom ops aren't supported).

**Comment by achalshah20** [comment]
@jackyko1991 

It might break existing eigen dependent app?. Shouldn't it be in include/tensorflow/external/ as these are tensorflow specific?

```
# Eigen directory
install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/eigen/src/eigen/Eigen/
        DESTINATION include/Eigen)
# external directory
install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/external/eigen_archive/
        DESTINATION include/external/eigen_archive)
# third_party eigen directory
install(DIRECTORY ${tensorflow_source_dir}/third_party/eigen3/
        DESTINATION include/third_party/eigen3)
# unsupported Eigen directory
install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/eigen/src/eigen/unsupported/Eigen/
        DESTINATION include/unsupported/Eigen)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gunan** [comment]
With dropped cmake support, and bazel officially working, what is the latest on this PR?

**Comment by jackyko1991** [comment]
It contains windows API release with reusable cmake packages.

I am not sure if bazel build on windows get GPU support if I want to code with MSVC.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jackyko1991** [comment]
@gunan I still see there are updates on cmake build, though latest release already changed to bazel build on windows.

TF cc core is packed as a single .dll with cmake config. With this users can link TF with their own applications.

The C++ api solution is only provided in linux (https://www.tensorflow.org/guide/extend/cc) and also hard for users to get all the necessary dependencies. I believe cmake contrib should not be removed as this PR would give a better C++ api experience in windows.

**Comment by gunan** [comment]
Cmake build is deprecated, and the cmakefiles are going to be removed together with the removal of  rest of contrib.

libtensorflow can be built on windows if you like to use the C++ API of TF. To pack things into a single DLL, you can simply use --config=monolithic with bazel build. So our bazel build is much more capable than the current cmake build. So I still do not understand what cmake build offers that bazel does not in terms of your requirements.

**Comment by jackyko1991** [comment]
So... should I close this PR?

**Comment by jackyko1991** [comment]
Btw will cmake TF build change to community supported? Most of cross platform C++ projects are on CMake, e.g. OpenCV, VTK, ITK. These are famous image processing toolkits and allow me to develop IO interfaces in specific domain application.

I am actually writing offline inference tools for medical image analysis, which greatly depends on cmake toolchains. But there is no way combing bazel output with cmake projects (e.g. header and dependency）. It is still worth to keep cmake.

**Comment by gunan** [comment]
You can reach out to `build@tensorflow.org` but the decision on cmake is unfortunately final.
There are volunteers in the mailing list that would like to keep cmake build working, but there is no official support for cmake going forward.

As for this pr, I am OK with accepting this, but I would like to see the changes in cmake folder and all the other changes (setup.py, C++ headers, etc) split to a different PR, to have better description on those. Because those changes have the potential to affect all of TF.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
